build-system/babel-plugins/babel-plugin-transform-log-methods/index.js

```diff
@@ -0,0 +1,288 @@
+ *    // transforms into 👇
+ *    dev().assert(foo != bar, ['1z'])
+ *
+ * `assert is overloaded so array syntax causes lookup of the message template
+ * and expansion of the message, or displaying a URL where the interpolated
+ * message is displayed.
+ *
+ * Additionally outputs a JSON file containing [template, id] pairs keyed by
+ * message template. This table is "backwards" for deduping, to key by id its
+ * rows should be reversed separately.
+ *
+ * When the template has arguments they are nested into the expansion method:
+ *
+ *    dev().assert(false, '%s should not be %s.', foo, bar)
+ *    // transforms into 👇
+ *    dev().assert(false, ['e2', foo, bar])
+ *
+ * It also nests arguments of template literals:
+ *
+ *    dev().assert(false, `Hello, my dear ${name}`);
+ *    // transforms into 👇
+ *    dev().assert(false, ['0a', name]);
+ *
+ * The motivation of this transform is to reduce binary size. The average length
+ * of a log message is ~43 whereas the common minified transformed output (as
+ * `['xx']`) is just 6.
+ *
+ * Resulting compressed, minified binaries reduce size by ~2.8% depending on
+ * their logging density.
+ */
+const base62 = require('base62/lib/ascii');
+const fs = require('fs');
+const path = require('path');
+const {
+  messagesPath,
+  singletonFunctions,
+  transformableMethods,
+} = require('../../log-module-metadata.js');
+
+/**
+ * Path to messages file for this version relative to the repo root.
+ * This is a default since tests can configure this table's output path, but for
+ * all building purposes this should be considered the canonical messages
+ * filepath for the built version.
+ */
+const defaultMessagesPath = messagesPath;
+
+/**
+ * Approximate length of a compressed message expansion call to determine
+ * whether an instance of message indirection actually reduces binary size.
+ */
+const roughMinifiedExpansionLength = '["xx"]'.length;
+
+const assertAliases = singletonFunctions.map(prefix => `${prefix}Assert`);
+
+/**
+ * Determines a filepath relative to the repo root.
+ * @param {string} path
+ * @return {string}
+ */
+const relativeToRoot = path => `${__dirname}/../../../${path}`;
+
+/**
+ * Reads the messages table from the JSON file.
+ * @param {string} messagesPath
+ * @return {!Object<string, string>}
+ */
+function getMessages(messagesPath) {
+  try {
+    return JSON.parse(fs.readFileSync(messagesPath));
+  } catch {
+    // When non-existent or empty just return an empty object. Transformations
+    // that read will write as well.
+    return {};
+  }
+}
+
+/**
+ * Writes the messages table to the JSON file.
+ * @param {string} messagesPath
+ * @param {!Object<string, string>} obj
+ */
+function writeMessages(messagesPath, obj) {
+  const json = JSON.stringify(obj, /* replacer */ null, /* spaces */ 2);
+  const messagesPathDir = path.dirname(messagesPath);
+  if (!fs.existsSync(messagesPathDir)) {
+    fs.mkdirSync(messagesPathDir);
```

Comment:
We should have a node module that makes directories recursively like `mkdir -p`

Reply:
`fs-extra`. So much nicer now, thanks!